File: basicFunctionCalls.js
import dotenv from 'dotenv';
import { ChatOpenAI } from "@langchain/openai";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";
import fs from 'fs/promises';

// Load environment variables
dotenv.config();

// Initialize the LLM with proper chat configuration
const llm = new ChatOpenAI({
    configuration: {
        baseURL: process.env.OPENROUTER_BASE_URL,
        defaultHeaders: {
            "HTTP-Referer": process.env.YOUR_SITE_URL,
            "X-Title": process.env.YOUR_SITE_NAME,
        }
    },
    modelName: process.env.OPENROUTER_MODEL_NAME,
    openAIApiKey: process.env.OPENROUTER_API_KEY,
});

// Define calculator tool
const calculate = async (mathExpression) => {
    try {
        console.log(`Calculator called with expressionnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn: ${mathExpression}`);
        
        // Strip whitespace and validate input
        mathExpression = mathExpression.trim();
        if (!mathExpression) {
            return "Error: Empty expression";
        }
        
        // Clean and validate expression
        const expression = mathExpression.replace(/[^\d+\-*/().\s]/g, '');
        const allowedChars = new Set("0123456789+-*/.()")
        if (![...expression.replace(/\s/g, '')].every(c => allowedChars.has(c))) {
            return "Error: Invalid characters in expression";
        }
        
        console.log(`Cleaned expression: ${expression}`);
        
        // Use Function constructor instead of eval for better security
        const result = new Function(`return ${expression}`)();
        console.log(`Calculation result: ${result}`);
        return result.toString();
    } catch (e) {
        console.log(`Calculator error: ${e.message}`);
        return `Error: ${e.message}`;
    }
};

// Define weather report tool
const getWeatherReport = async (location) => {
    try {
        console.log(`Weather report requested for: ${location}`);
        console.log(`Weather report requestedSDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD for: ${location}`);

        // Dummy weather data
        const weatherConditions = [
            'sudasdasnny', 'partlyasdasdasd clasdasdoudy', 'clasddasdasasdoudy', 'raiasdasny', 'stoasdasrmy', 'snowasdasdy'
        ];
        const temperatures = {
            celsius: Math.floor(Math.random() * 35) + 5,  // Random temp between 5-40°C
        };
        temperatures.fahrenheit = Math.round(temperatures.celsius * 9/5 + 32);
        
        const humidity = Math.floor(Math.random() * 60) + 30;  // Random humidity 30-90%
        const condition = weatherConditions[Math.floor(Math.random() * weatherConditions.length)];
        
        return `Weather report for ${location}:
Temperature: ${temperatures.celsius}°C (${temperatures.fahrenheit}°F)
Condition: ${condition}
Humidity: ${humidity}%`;
    } catch (e) {
        console.log(`Weather report error: ${e.message}`);
        return `Error getting weather report: ${e.message}`;
    }
};

const findExistingFunction = async (fnName) => {
    const files = await fs.readdir('FUNCTION');
    // Look for exact function name match first, ignoring file extensions
    return files.find(file => 
        file.toLowerCase().replace(/\.[^/.]+$/, "") === fnName.toLowerCase() ||
        file.toLowerCase().includes(fnName.toLowerCase())
    );
};

const createAndStoreFunction = async (functionDefinition) => {
    try {
        console.log('🔍 Checking function existence...');
        
        // Extract function name
        const fnNameMatch = functionDefinition.match(/function\s+(\w+)/);
        if (!fnNameMatch) {
            console.log('❌ Error: Invalid function definition format');
            return null;
        }
        const fnName = fnNameMatch[1];
        
        // Create FUNCTION directory if it doesn't exist
        await fs.mkdir('FUNCTION', { recursive: true });

        // Check for existing function
        const existingFile = await findExistingFunction(fnName);
        
        if (existingFile) {
            console.log('📌 Found existing function:', existingFile);
            const existingCode = await fs.readFile(`FUNCTION/${existingFile}`, 'utf8');
            console.log('♻️ Reusing existing function - no new file created');
            return new Function(`return (${existingCode})`)();
        }

        // Only proceed with creation if no existing function was found
        console.log('🆕 No existing function found. Creating new one...');
        
        // Add console.log to the function definition
        const enhancedFunction = functionDefinition.replace(
            /function\s+(\w+)\s*\((.*?)\)\s*{/,
            `function $1($2) {
    console.log('📞 Calling function: $1');
    console.log('📎 Arguments:', $2);`
        );

        // Generate filename based on function name
        const filename = `${fnName.toLowerCase()}.txt`;
        const filepath = `FUNCTION/${filename}`;
        
        await fs.writeFile(filepath, enhancedFunction, 'utf8');

        console.log('✅ New function created successfully');
        console.log('💾 Stored as:', filename);
        
        return new Function(`return (${enhancedFunction})`)();
    } catch (e) {
        console.error('❌ Error in function creation process:', e.message);
        return null;
    }
};

// Modified agent function to be more strict about function reuse
const agentNode = async (state) => {
    console.log('\n🤖 Agent Node Started');
    const systemPrompt = `You are a helpful assistant that can create and execute custom functions.
    When you need to create a function, write it in the following format:
    <function>
    function customFunction(params) {
        // function body
        return result;
    }
    </function>
    
    Then call it using:
    <execute>customFunction(args)</execute>
    `;
    
    const messages = [
        new SystemMessage(systemPrompt),
        new HumanMessage(state.input)
    ];
    
    const response = await llm.invoke(messages);
    let result = response.content;
    
    // Extract and process function definitions
    const functionPattern = /<function>([\s\S]*?)<\/function>/g;
    const functionMatches = result.matchAll(functionPattern);
    const functions = new Map();
    
    for (const match of functionMatches) {
        const functionDef = match[1].trim();
        const fnNameMatch = functionDef.match(/function\s+(\w+)/);
        if (fnNameMatch) {
            const fnName = fnNameMatch[1];
            console.log(`\n🔄 Processing function: ${fnName}`);
            const fn = await createAndStoreFunction(functionDef);
            if (fn) {
                functions.set(fnName, fn);
                result = result.replace(match[0], `Using function: ${fnName}`);
            }
        }
    }
    
    // Execute function calls
    const executePattern = /<execute>(.*?)<\/execute>/g;
    const executeMatches = result.matchAll(executePattern);
    
    for (const match of executeMatches) {
        console.log('\n⚡ Executing function call');
        const call = match[1].trim();
        const fnName = call.match(/(\w+)\(/)[1];
        const fn = functions.get(fnName);
        
        if (fn) {
            try {
                const argsMatch = call.match(/\((.*)\)/)[1];
                const args = eval(`[${argsMatch}]`);
                console.log('📊 Function:', fnName);
                console.log('🔢 Arguments:', args);
                console.log('🎬 Starting execution...');
                const execResult = await fn(...args);
                console.log('✨ Result:', execResult);
                console.log('✅ Execution completed');
                result = result.replace(match[0], String(execResult));
            } catch (e) {
                console.error('❌ Execution error:', e.message);
                result = result.replace(match[0], `Error executing function: ${e.message}`);
            }
        }
    }
    
    return { output: result };
};

// Test function
const runTests = async () => {
    const testCases = [
        `Create a function to calculate the area of a circle and use it:
        <function>
        function calculateCircleArea(radius) {
            return Math.PI * radius * radius;
        }
        </function>
        The area of a circle with radius 5 is: <execute>calculateCircleArea(5)</execute>
        `
    ];

    try {
        for (const query of testCases) {
            const state = { input: query, output: null };
            const result = await agentNode(state);
            console.log(`\nQuery: ${query}`);
            console.log(`Result: ${result.output}`);
        }
    } catch (e) {
        console.error(`An error occurred: ${e.message}`);
    }
};

// Run the tests
runTests(); 

File: basicFunctionCalls.py
import os
from typing import TypedDict, Annotated, Sequence
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage
from langgraph.graph import StateGraph, END
from langchain_core.tools import tool
import re
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Initialize the LLM with proper chat configuration
llm = ChatOpenAI(
    base_url=os.getenv('OPENROUTER_BASE_URL'),
    api_key=os.getenv('OPENROUTER_API_KEY'),
    model=os.getenv('OPENROUTER_MODEL_NAME'),
    # temperature=0
)

# Define state structure
class AgentState(TypedDict):
    input: str
    output: str | None

# Define calculator as a tool using the new @tool decorator
@tool
def calculate(math_expression: str) -> str:
    """Useful for mathematical calculations. Input should be a valid numerical expression using +, -, *, /, or parentheses."""
    try:
        print(f"Calculator called with expression: {math_expression}")  # Better debug message
        print("Calculator called with expressionnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn: ", math_expression)
        # Strip whitespace and validate input
        math_expression = math_expression.strip()
        if not math_expression:
            return "Error: Empty expression"
            
        expression = re.sub(r'[^\d+\-*/().\s]', '', math_expression)
        allowed_chars = set("0123456789+-*/.()")
        if not all(c in allowed_chars for c in expression.replace(" ", "")):
            return "Error: Invalid characters in expression"
            
        # Additional debug
        print(f"Cleaned expression: {expression}")
        
        result = eval(expression)
        print(f"Calculation result: {result}")  # Debug the result
        return str(result)
    except Exception as e:
        print(f"Calculator error: {str(e)}")  # Debug any errors
        return f"Error: {str(e)}"

# Define the agent function
def agent_node(state: AgentState, tools: Sequence[tool]):
    # Create a prompt that includes tool information
    system_prompt = """You are a helpful assistant with access to the following tools:
    - calculate: Use this for mathematical calculations. Provide numerical expressions using +, -, *, /, or parentheses.

    When you need to perform calculations, use the calculate tool by writing:
    <tool>calculate|your expression here</tool>
    """
    
    messages = [
        SystemMessage(content=system_prompt),
        HumanMessage(content=state["input"])
    ]
    
    response = llm.invoke(messages)
    
    # Extract tool calls using regex
    tool_pattern = r'<tool>calculate\|(.*?)</tool>'
    matches = re.finditer(tool_pattern, response.content)
    
    # Process any tool calls
    result = response.content
    for match in matches:
        expression = match.group(1)
        calculation_result = calculate(expression)
        result = result.replace(match.group(0), calculation_result)
    
    state["output"] = result
    return {"output": state["output"]}

# Create the graph
workflow = StateGraph(AgentState)

# Add the agent node
workflow.add_node("agent", lambda state: agent_node(state, [calculate]))

# Define the edges
workflow.set_entry_point("agent")
workflow.add_edge("agent", END)

# Compile the graph
app = workflow.compile()

# Test queries
test_cases = [
    "who is the president of the united states?",
    "Calculate (12345 * 6789) / (12 + 34.5)",
    "What is 15 multiplied by 3?"
]

# Run test cases
try:
    for query in test_cases:
        result = app.invoke({"input": query, "output": None})
        print(f"\nQuery: {query}")
        print(f"Result: {result['output']}")
except Exception as e:
    print(f"An error occurred: {str(e)}")